README.md

```diff
@@ -32,59 +37,224 @@ The main features of NEMU include
+### Install dependencies
+
+Because different distributions have different package management tools, the installation commands are different.
+For Ubuntu, users can install the dependencies with the following command:
+```bash
+apt install build-essential man gcc gdb git libreadline-dev libsdl2-dev zstd libzstd-dev
+```
+
+### Use NEMU as reference design
+
+#### Build reference.so
+
+To build NEMU as reference design, run
+``` bash
+make xxx-ref_defconfig
+make menuconfig  # save
+make -j
+```
+`./build/riscv64-nemu-interpreter-so` is the reference design.
+
+Specifically, xxx-ref_defconfig varies for different ISA extensions.
+
+| rv64gcb | rv64gcbh | rv64gcbv |
+| :-----: | :------: | :-------: |
+|  riscv64-xs-ref_defconfig | riscv64-rvh-ref_defconfig | riscv64-rvv-ref_defconfig |
+
+
+#### Cosimulation
+
+To test XS-GEM5 against NEMU, refer to [the doc of XS-GEM5 Difftest](https://github.com/OpenXiangShan/GEM5?tab=readme-ov-file#difftest-with-nemu).
+
+To test Xiangshan processor against NEMU, run
+``` bash
+./emu \
+  -i test_workload.bin \
+  --diff $NEMU_HOME/build/riscv64-nemu-interpreter-so \
+  2> perf.out
 ```
+Details can be found in [the tutorial of Xiangshan](https://xiangshan-doc.readthedocs.io/zh-cn/latest/tutorials/rvsc23/).
+
+### Workloads
+
+As described in [the workflow](#run-in-standalone-mode-without-checkpoint), NEMU either takes a baremetal app or
+an operating system image as input.
+
+For baremetal app, [Abstract Machine](https://github.com/OpenXiangShan/nexus-am) is a light-weight baremetal library.
+Common simple apps like coremark and dhrystone can be built with Abstract Machine.
+
+For build operating system image,
+Please read [the doc to build Linux](https://xiangshan-doc.readthedocs.io/zh-cn/latest/tools/linux-kernel-for-xs/).
+
+Then modify `NEMU_HOME` and `BBL_PATH` in `$NEMU_HOME/scripts/checkpoint_example/checkpoint_env.sh` and the workload parameter passed to the function in each example script to get started.
+
+
+### SimPoint profiling and checkpoint
+
+Please read [the doc to generate checkpoint](https://xiangshan-doc.readthedocs.io/zh-cn/latest/tools/simpoint/)
+
+### Run a checkpoint with XS-GEM5 or Xiangshan processor
+
+Run a checkpoint with Xiangshan processor
+
+```bash
+./build/emu -i /path/to/a/checkpoint.gz
+```
+
+Run checkpoints with XS-GEM5: [the doc to run XS-GEM5](https://github.com/OpenXiangShan/GEM5?tab=readme-ov-file#run-gem5)
+
+## FAQ
+
+### Why cannot produce a checkpoint in M-mode?
+
+Read the source code of [GCPT restorer](https://github.com/OpenXiangShan/NEMU/blob/master/resource/gcpt_restore/src/restore.S)
+
+Because we restore checkpoint in M mode, and the PC of returning to user mode is stored in EPC register.
+This recovery method will break the architecture state (EPC) if the checkpoint is produced in M mode.
+In contrast, if the checkpoint is produced in S mode or U mode,
+the return process is just like a normal interrupt return, which will not break the architecture state.
```

Comment:
like a normal trap return

trap is more suitable here for both exceptions and interrupts